lib/bit-representation.nix

```
@@ -0,0 +1,106 @@
+/* Collection of functions for bit representation conversions such as binary,
+   octal, decimal and hexadecimal.
+
+   These functions works by using a bit mapping, a attrset that maps each
+   character to a numerical value. The base of the numeric system is the length
+   of the keys of this attrset.
+*/
+
+{ lib }:
+
+let
+    remainder = x: y: x - ((builtins.div x y)*y);
+    invertAttrset = attrs: let
+      keys = lib.attrNames attrs;
+    in lib.foldl (x: y: x // {
+      "${toString attrs.${y}}" = y;
+    }) {} keys;
+in {
+  /* Default bit mappings for the most common bases */
+  bitMappings = rec { # maps characters to values
+    bin = {
+      "0" = 0;
+      "1" = 1;
+    };
+    oct = bin // {
+      "2" = 2;
+      "3" = 3;
+      "4" = 4;
+      "5" = 5;
+      "6" = 6;
+      "7" = 7;
+    };
+    dec = oct // {
+      "8" = 8;
+      "9" = 9;
+    };
+    hex = dec // {
+      a = 10;
+      b = 11;
+      c = 12;
+      d = 13;
+      e = 14;
+      f = 15;
+    };
+  };
```

Since this is a read-only constant mapping, why not unroll it instead of using an otherwise useless `rec`?

```suggestion
  bitMappings = { # maps characters to values
    binary = {
      "0" = 0;
      "1" = 1;
    };
    octal = {
      "0" = 0;
      "1" = 1;
      "2" = 2;
      "3" = 3;
      "4" = 4;
      "5" = 5;
      "6" = 6;
      "7" = 7;
    };
    decimal = {
      "0" = 0;
      "1" = 1;
      "2" = 2;
      "3" = 3;
      "4" = 4;
      "5" = 5;
      "6" = 6;
      "7" = 7;
      "8" = 8;
      "9" = 9;
    };
    hexadecimal = {
      "0" = 0;
      "1" = 1;
      "2" = 2;
      "3" = 3;
      "4" = 4;
      "5" = 5;
      "6" = 6;
      "7" = 7;
      "8" = 8;
      "9" = 9;
      a = 10;
      b = 11;
      c = 12;
      d = 13;
      e = 14;
      f = 15;
    };
  };
```

Also, what about using uppercase letters to the hexadecimal extras?

---

It can be standarized to be uppercase. Right now I am lowercasing all text before splitting into chars.

BTW why not keep it this way? It's shorter and the bigger charsets are just supersets of the smaller ones.

---

It can be standarized to be uppercase. Right now I am lowercasing all text before splitting into chars.

BTW why not keep it this way? It's shorter and the bigger charsets are just supersets of the smaller ones.

---

Yeah, I think I like the expanding approach.